Added support to download nuget and added python code to automate the…

angelortiz1007 opened the pull request:
… x64 and arm64 mach build -r/-d --uwp build process.

<!-- Please describe your changes on the following line: -->
Modified/added python code to:
1. Download nuget which is required for building support\HoloLens
2. added python code to perform a nuget install/restore ServoApp.sln which causes the Angle libraries to be installed so that mach build will succeed.
3. added python code to perform a MsBuild of the support\HoloLens\ServoApp.sln solution so that mach build will build ServoApp and create a ServoApp .appx install file.

Note: The Msbuild path must be present otherwise the msbuild command will fail.


---
<!-- Thank you for contributing to Servo! Please replace each `[ ]` by `[X]` when the step is complete, and replace `___` with appropriate data: -->
- [x] `./mach build -d` does not report any errors
- [x] `./mach test-tidy` does not report any errors
- [ ] These changes fix #23753 (GitHub issue number if applicable)

<!-- Either: -->
- [x] There are tests for these changes OR
- [x] These changes do not require tests because by performing python mach build -r/-d --uwp ---win-arm64 or mach build -r/-d --uwp

<!-- Also, please make sure that "Allow edits from maintainers" checkbox is checked, so that we can help you if you get stuck somewhere along the way.-->

<!-- Pull requests that do not address these steps are welcome, but they will require additional verification as part of the review process. -->

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/servo/servo/23961)
<!-- Reviewable:end -->

angelortiz1007 committed:
Added support to download nuget and added python code to automate the x64 and arm64 mach build -r/-d --uwp build process.

angelortiz1007 commented:
Important note.
When building, the msbuild.exe must be within the %PATH%.
If you open a x86_64 VS2017 command prompt msbuild will be found.

if you perform a cross platform build by opening a regular command prompt, you must set the path to where msbuild.exe is prior to:
1. set PKG_CONFIG_ALLOW_CROSS=1
2. python mach build -r --uwp --win-arm64

**Here is an attempt to determine the msbuild.exe path at build time.**
**VSINSTALLDIR**=C:\Program Files (x86)\Microsoft Visual Studio\2017\Professional\

location of Msbuild.exe => MSBuild\15.0\Bin

**VisualStudioVersion** =15.0

set **MSBUILDDIR**=%VSINSTALLDIR%MsBuild\%VisualStudioVersion%\bin

nirbheek commented:
> Here is an attempt to determine the msbuild.exe path at build time.

The approach used in GStreamer's Cerbero build system is to extract `PATH` from the environment set by `vcvarsall.bat`: https://gitlab.freedesktop.org/gstreamer/cerbero/blob/master/cerbero/ide/vs/env.py#L106

Might be worth trying to do the same in Servo too.

angelortiz1007 commented:
> Here is an attempt to determine the msbuild.exe path at build time.
> 
> The approach used in GStreamer's Cerbero build system is to extract PATH from the environment set by vcvarsall.bat: https://gitlab.freedesktop.org/gstreamer/cerbero/blob/master/cerbero/ide/vs/env.py#L106
> Might be worth trying to do the same in Servo too.

Thank you for looking at the pull request and providing this suggestion.  I will review your suggestion!

Thank you,
Angel

jdm commented on python/servo/build_commands.py:
```diff
@@ -289,6 +289,10 @@ def build(self, release=False, dev=False, jobs=None, params=None,
 
+            print("*** calling angle_root() ***")
+            angle_root(target_triple)
+            print("*** angle_root() completed ***")
```
These print lines should be removed.

angelortiz1007 replied:
Done.

jdm commented on python/servo/build_commands.py:
```diff
@@ -764,6 +772,34 @@ def gstreamer_root(target, env):
 
+def angle_root(target):
+    arch = {
+
```
nit: remove this blank line.

angelortiz1007 replied:
Done.

jdm replied:
It's still there :)

jdm commented on python/servo/build_commands.py:
```diff
@@ -764,6 +772,34 @@ def gstreamer_root(target, env):
+    print("angle_x64 Architecture = {}".format(angle_x64))
+    angle_default_path = path.join(os.getcwd(),
+                                   "support\\hololens\\packages\\ANGLE.WindowsStore.Servo.2.1.13\\bin\\UAP\\")
+    angle_default_path += angle_x64
```
We already have code to create this path in https://github.com/servo/servo/blob/91469aa47215c8e1b4e4ae5643ca793b3f65af9b/python/servo/build_commands.py#L295-L298, so let's use that code in this method (since the use of path.join there is a bit easier to read) and call angle_root in that code to avoid duplication.

angelortiz1007 replied:
Done.

jdm commented on python/servo/build_commands.py:
```diff
@@ -764,6 +772,34 @@ def gstreamer_root(target, env):
+        print("Angle Found...")
+    else:
+        print("Angle not found.  Performing nuget to restore Angle package")
+        os.system(nuget_command)
```
Rather than using os.system directly, let's use the `check_call` helper like other parts of this file: https://github.com/servo/servo/blob/91469aa47215c8e1b4e4ae5643ca793b3f65af9b/python/servo/build_commands.py#L234-L235

jdm commented on python/servo/build_commands.py:
```diff
@@ -764,6 +772,34 @@ def gstreamer_root(target, env):
+    angle_default_path += angle_x64
+
+    # Nuget executable command
+    nuget_exe = path.join(os.getcwd(), ".servo\\msvc-dependencies\\nuget\\08-08-2019\\nuget.exe ")
```
We should use `path.join(self.msvc_package_dir('nuget'), 'nuget.exe')` instead of hardcoding this (https://github.com/servo/servo/blob/91469aa47215c8e1b4e4ae5643ca793b3f65af9b/python/servo/command_base.py#L579-L580)

angelortiz1007 replied:
Done.

jdm commented on python/servo/build_commands.py:
```diff
@@ -764,6 +772,34 @@ def gstreamer_root(target, env):
+    # Nuget executable command
+    nuget_exe = path.join(os.getcwd(), ".servo\\msvc-dependencies\\nuget\\08-08-2019\\nuget.exe ")
+    nuget_action = "restore "
+    nuget_app = path.join(os.getcwd(), "support\\hololens\\ServoApp.sln")
```
Let's use `path.join(os.getcwd(), "support", "hololens", "ServoApp.sln")`

jdm commented on python/servo/build_commands.py:
```diff
@@ -289,6 +289,10 @@ def build(self, release=False, dev=False, jobs=None, params=None,
                 sys.exit(1)
 
+            print("*** calling angle_root() ***")
+            angle_root(target_triple)
```
The return value can be used just below this to set the LIB variable.

angelortiz1007 replied:
Done.

jdm commented on python/servo/build_commands.py:
```diff
@@ -905,6 +941,33 @@ def package_gstreamer_dlls(env, servo_exe_dir, target, uwp):
+    # msbuild /m:4 /p:project=ServoApp .\support\hololens\servoapp.sln /p:SolutionDir=.\support\hololens
+    # /p:Configuration="Debug" /p:Platform="x64" /property:AppxBundle=Always;AppxBundlePlatforms="x64"
+
+    msbuild_cmd = "msbuild /m:4 /p:project=ServoApp .\support\hololens\servoapp.sln /p:SolutionDir=.\support\hololens "
```
Let's use `/m` to avoid hardcoding a particular number.

jdm commented on python/servo/build_commands.py:
```diff
@@ -905,6 +941,33 @@ def package_gstreamer_dlls(env, servo_exe_dir, target, uwp):
+    msbuild_cmd = "msbuild /m:4 /p:project=ServoApp .\support\hololens\servoapp.sln /p:SolutionDir=.\support\hololens "
+    msbuild_conf = "/p:Configuration=" + Configuration + " /p:Platform=" + vs_platform + \
+        " /p:AppxBundle=Always;AppxBundlePlatforms=" + vs_platform
+    os.system(msbuild_cmd + msbuild_conf)
```
We should use check_call here as well.

angelortiz1007 replied:
Done.

jdm commented on python/servo/build_commands.py:
```diff
@@ -701,6 +705,10 @@ def package_generated_shared_libraries(libs, build_path, servo_exe_dir):
 
+                # UWP build hololens
+                if uwp:
+                    build_uwp_hololens(target_triple, dev)
```
We will want to check the return value here and set status to 1 if the build process fails.

jdm reviewed (changes requested):
Looks like a good start! I've left some comments for ways to make the code a bit more idiomatic and integrate better with the rest of the build system.

jdm commented:
@bors-servo try=windows

bors-servo commented:
:hourglass: Trying commit d28df220f17b9a601b939d6af9630f44491d9c47 with merge 55e471f6e8bc3be5f01908514917be2fd728c8ad...

bors-servo commented:
:broken_heart: Test failed - [status-taskcluster](https://tools.taskcluster.net/task-group-inspector/#/PllQrqRYTJ69Us27xND8bA)

jdm commented:
@bors-servo retry

bors-servo commented:
:hourglass: Trying commit d28df220f17b9a601b939d6af9630f44491d9c47 with merge 227ad85b4edca71d0216a4070f745f925db7a830...

jdm commented on python/servo/build_commands.py:
```diff
@@ -764,6 +765,33 @@ def gstreamer_root(target, env):
+                                   "ANGLE.WindowsStore.Servo.2.1.13", "bin", "UAP", angle_arch)
+
+    # Nuget executable command
+    nuget_exe = path.join(self.msvc_package_dir('nuget'), 'nuget.exe ')
```
Actually, given that we add the nuget directory to the PATH, we should just be able to refer to `nuget.exe` directly without using `self.msvc_package_dir`. That means we can remove the self argument from this function as well.

angelortiz1007 replied:
Humm… I must be misunderstanging….

If I remove the nuget_exe = path.join(self.msvc_package_dir('nuget'), 'nuget.exe ') line and just refer to 'nuget.exe'  I receive the following error message.  Seems the location of nuget.exe can't be found.

I get the following error:
CalledProcessError: Command 'n u g e t . e x e   r e s t o r e   C : \ g i t - m o z i l l a \ a n g e l o r t i z 1 0 0 7 - 2 \ s u p p o r t \ h o l o l e n s \ S e r v o A p p . s l n' returned non-zero exit status 1

  File "C:\git-mozilla\angelortiz1007-2\python\servo\build_commands.py", line 294, in build
    append_to_path_env(angle_root(target_triple),
  File "C:\git-mozilla\angelortiz1007-2\python\servo\build_commands.py", line 790, in angle_root
    check_call(nuget_command)
  File "C:\git-mozilla\angelortiz1007-2\python\servo\command_base.py", line 195, in check_call
    raise subprocess.CalledProcessError(status, ' '.join(*args))

jdm replied:
That happens because you are still passing a string to check_call, and not a list like I proposed in my other review comments.

jdm replied:
Testing some related changes locally, it looks like you will need `check_call(["nuget", ...], env=env)` to allow the nuget executable to be found in the path.

angelortiz1007 replied:
Josh, sorry I must be missing something.

I've got:
    if not os.path.exists(angle_default_path):
        check_call(['nuget.exe', 'restore', nuget_app], env)

env=env too.

I get the following error:

TypeError: bufsize must be an integer

  File "C:\git-mozilla\angelortiz1007-2\python\servo\build_commands.py", line 294, in build
    append_to_path_env(angle_root(target_triple, env), env, "LIB")
  File "C:\git-mozilla\angelortiz1007-2\python\servo\build_commands.py", line 779, in angle_root
    check_call(['nuget.exe', 'restore', nuget_app], env)
  File "C:\git-mozilla\angelortiz1007-2\python\servo\command_base.py", line 182, in check_call
    proc = subprocess.Popen(*args, shell=sys.platform == 'win32', **kwargs)
  File "C:\Python27\lib\subprocess.py", line 343, in __init__
    raise TypeError("bufsize must be an integer")

jdm replied:
I suspect that you do not actually have a variable named `env` at the time you are calling check_call. I recommend passing it in to angle_root as an argument; it should be available in the caller.

angelortiz1007 replied:
Yep, I'm passing it in.  Here is my function signature...
def angle_root(target, env):
    arch = {

Here is my call to angle_root(…):
append_to_path_env(angle_root(target_triple, env), env, "LIB")

Something is missing.... Maybe in....
windows_build_env = {
    "x86_64": {
        "GSTREAMER_1_0_ROOT_X86_64": "%HOMEDRIVE%%HOMEPATH%\\gst\\gstreamer\\1.0\\x86_64\\",
    },
    "arm64": {
        "PKG_CONFIG_ALLOW_CROSS": "1",
    },
    "all": {
        "PYTHON3": "%HOMEDRIVE%%HOMEPATH%\\python3\\python.exe",
        "LINKER": "lld-link.exe",
    },
}

jdm replied:
So in general, the "bufsize must be an integer" error occurs when the parameters to check_call are not being passed correctly (eg. https://mihirknows.blogspot.com/2011/02/solution-raise-typeerrorbufsize-must-be.html, https://stackoverflow.com/questions/34156193/subprocess-cp-returns-error-bufsize-must-be-integer/34156248, https://stackoverflow.com/questions/17824096/bufsize-must-be-an-integer-error-while-grepping-a-message).

angelortiz1007 replied:
Done.

jdm commented on python/servo/build_commands.py:
```diff
@@ -764,6 +765,33 @@ def gstreamer_root(target, env):
+        print("Angle Found...")
+    else:
+        print("Angle not found.  Performing nuget to restore Angle package")
+        check_call(nuget_command)
```
Instead of passing a single string, we should use a list:
```py
check_call(["nuget", "restore", nuget_app])
```

jdm commented on python/servo/build_commands.py:
```diff
@@ -905,6 +933,34 @@ def package_gstreamer_dlls(env, servo_exe_dir, target, uwp):
+    msbuild_cmd = "msbuild /m /p:project=ServoApp .\support\hololens\servoapp.sln /p:SolutionDir=.\support\hololens "
+    msbuild_conf = "/p:Configuration=" + Configuration + " /p:Platform=" + vs_platform + \
+        " /p:AppxBundle=Always;AppxBundlePlatforms=" + vs_platform
+    check_call(msbuild_cmd + msbuild_conf)
```
We should also pass a list instead of a string.

angelortiz1007 replied:
Having a little bit of an issue here. If I change check_call() to:
**check_call([msbuild_cmd, msbuild_conf])**
I get the following error.  It seems line 195 has self.config["build"]["mode"] returning a ""/blank.
CalledProcessError: Command 'msbuild /m /p:project=ServoApp .\support\hololens\servoapp.sln /p:SolutionDir=.\support\hololens  /p:Configuration=Release /p:Platform=x64 /p:AppxBundle=Always;AppxBundlePlatforms=x64' returned non-zero exit status 1

  File "C:\git-mozilla\angelortiz1007-2\python\servo\build_commands.py", line 704, in build
    if build_uwp_hololens(target_triple, dev):
  File "C:\git-mozilla\angelortiz1007-2\python\servo\build_commands.py", line 961, in build_uwp_hololens
    check_call([msbuild_cmd, msbuild_conf])
  File "C:\git-mozilla\angelortiz1007-2\python\servo\command_base.py", line 195, in check_call
    raise subprocess.CalledProcessError(status, ' '.join(*args))

If I change check_call([…]) can convert the msbuild_cmd string and msbuild_conf string into a list, msbuild fails because it appears a " " is added after /p:Platform= x64 and /p:Configuration= Release.

For msbuild to work, there can't be spaces between the "=".

Any clarification or suggestions?

jdm replied:
Each individual argument should be its own list item: `["msbuild", "/m", "/p:project=ServoApp", ...]`

angelortiz1007 replied:
Ok,  so I did split things up into individual list items "[…].  However, when the final command is constructed and executed, the list items are separated by a space.  This is normally a good thing.  However, in 3 of the msbuild parameters, there can't be a space between the command line option and it's value.  E.G.

/p:Configuration=Release will work, but /p:Configuration= Release will not.  I did a hybrid below in order to get msbuild to work.  Can you tell me what am I missing?

def build_uwp_hololens(target, dev):
    # determine Visual studio Build Configuration (Debug/Release) and
    # Build Platform (x64 vs arm64)
    vs_platforms = {
        "x86_64": "x64",
        "i686": "x86",
        "aarch64": "arm64",
    }
    target_arch = target.split('-')[0]
    vs_platform = vs_platforms[target_arch]

    if dev:
        Configuration = "Debug"
    else:
        Configuration = "Release"

    # execute msbuild
    # Note: /m implies to use as many CPU cores as possible while building.
    # msbuild /m /p:project=ServoApp .\support\hololens\servoapp.sln /p:SolutionDir=.\support\hololens
    # /p:Configuration="Debug" /p:Platform="x64" /property:AppxBundle=Always;AppxBundlePlatforms="x64"
    check_call(["msbuild", "/m", "/p:project=ServoApp", ".\support\hololens\servoapp.sln", **"/p:SolutionDir=.\support\hololens"**,
                **'/p:Configuration=' + Configuration**, **"/p:Platform=" + vs_platform** ])

jdm replied:
That looks right - any values that need to be a single string should, in fact, be a single string :)

angelortiz1007 replied:
Done.

jdm commented on python/servo/build_commands.py:
```diff
@@ -764,6 +765,33 @@ def gstreamer_root(target, env):
+        "x86_64": "x64",
+    }
+    angle_arch = arch[target.split('-')[0]]
+    print("angle_x64 Architecture = {}".format(angle_arch))
```
No need for this print.

angelortiz1007 replied:
Done.

jdm commented on python/servo/build_commands.py:
```diff
@@ -764,6 +765,33 @@ def gstreamer_root(target, env):
+
+    if os.path.exists(angle_default_path):
+        print("Angle Found...")
+    else:
```
Let's do:
```py
if not os.path.exists(angle_default_path):
    ...
```
And remove the print statements.

jdm commented on python/servo/build_commands.py:
```diff
@@ -701,6 +697,11 @@ def package_generated_shared_libraries(libs, build_path, servo_exe_dir):
 
+                # UWP build hololens
+                if uwp:
+                    if build_uwp_hololens(target_triple, dev):
```
I was mistaken earlier - check_call will automatically cause the build to fail if it encounters a problem, so we don't need to check the return value and set status to 1.

jdm commented on python/servo/build_commands.py:
```diff
@@ -291,14 +291,10 @@ def build(self, release=False, dev=False, jobs=None, params=None,
+            append_to_path_env(angle_root(self, target_triple),
+                               env,
+                               "LIB"
+                               )
```
nit: this can not fit on one line - `append_to_path_env(angle_root(target_triple), env, "LIB")`

angelortiz1007 commented:
Implemented changes requested by JDM and performed the below tests.  Rebased changes and performed a pull request for testing.

Ran tests to verify the following passed:
1. mach build -r/-d --uwp
2. python mach build -r --uwp --win-arm64
3. Tested return value from build_uwp_hololens() which calls check_call() by modifying msbuild parameter .\support\hololens\servoapp.sln to servoapp.sln2.
Upon performing a build the logic introduced to set status = 1was tested.  Output from test is shown below:
MSBUILD : error MSB1009: Project file does not exist.
Switch: .\support\hololens\servoapp.sln2
Error running mach:

    ['build', '-r', '--uwp']

The error occurred in code that was called by the mach command. This is either
a bug in the called code itself or in the way that mach is calling it.

You should consider filing a bug for this issue.

If filing a bug, please include the full output of mach, including this error
message.

The details of the failure are as follows:

CalledProcessError: Command 'm s b u i l d   / m   / p : p r o j e c t = S e r v o A p p   . \ s u p p o r t \ h o l o l e n s \ s e r v o a p p . s l n 2   / p : S o l u t i o n D i r = . \ s u p p o r t \ h o l o l e n s   / p : C o n f i g u r a t i o n = R e l e a s e   / p : P l a t f o r m = x 6 4   / p : A p p x B u n d l e = A l w a y s ; A p p x B u n d l e P l a t f o r m s = x 6 4' returned non-zero exit status 1

  File "C:\git-mozilla\angelortiz1007-2\python\servo\build_commands.py", line 702, in build
    if build_uwp_hololens(target_triple, dev):
  File "C:\git-mozilla\angelortiz1007-2\python\servo\build_commands.py", line 962, in build_uwp_hololens
    check_call(msbuild_cmd + msbuild_conf)
  File "C:\git-mozilla\angelortiz1007-2\python\servo\command_base.py", line 195, in check_call
    raise subprocess.CalledProcessError(status, ' '.join(*args))

jdm commented:
@bors-servo try=windows

bors-servo commented:
:hourglass: Trying commit cd6b048a51c3a8ced53db5a1416cd9eb86d26acd with merge cad7e839086f82779876acde7a17e13b6f36c0d5...

bors-servo commented:
:broken_heart: Test failed - [status-taskcluster](https://tools.taskcluster.net/task-group-inspector/#/X73VYqukQoWl1y2kldNO2w)

jdm commented:
```
Error running mach:

    ['build', '--dev', '--uwp', '--win-arm64']

The error occurred in code that was called by the mach command. This is either
a bug in the called code itself or in the way that mach is calling it.

You should consider filing a bug for this issue.

If filing a bug, please include the full output of mach, including this error
message.

The details of the failure are as follows:

CalledProcessError: Command 'msbuild /m /p:project=ServoApp .\support\hololens\ServoApp.sln /p:SolutionDir=.\support\hololens /p:Configuration=Debug /p:Platform=arm64 /p:AppxBundle=Always;AppxBundlePlatforms=arm64' returned non-zero exit status 1

  File "C:\Users\task_1566239357\repo\python\servo\build_commands.py", line 705, in build
    build_uwp_hololens(target_triple, dev)
  File "C:\Users\task_1566239357\repo\python\servo\build_commands.py", line 976, in build_uwp_hololens
    "/p:AppxBundle=Always;AppxBundlePlatforms=" + vs_platform])
  File "C:\Users\task_1566239357\repo\python\servo\command_base.py", line 195, in check_call
    raise subprocess.CalledProcessError(status, ' '.join(*args))
```
Let's try using `check_output` instead of check_call, which should give us any output from the failing msbuild command.

jdm commented:
Oh, actually, I bet the problem is that it can't find msbuild in the arm64 build configuration because we have to run it from outside of the VS command shell. We need to find the path to the msbuild binary manually like [this code](https://github.com/servo/servo/blob/a084997afee23bb541e89a807905ff1c815a649e/python/servo/build_commands.py#L253-L269).

angelortiz1007 commented:
> Oh, actually, I bet the problem is that it can't find msbuild in the arm64 build configuration because we have to run it from outside of the VS command shell. We need to find the path to the msbuild binary manually like this code.

Josh, you are correct.  msbuild is not found in this case.  

When I build it locally, I have to add msbuild to my path.

angelortiz1007 commented:
> Oh, actually, I bet the problem is that it can't find msbuild in the arm64 build configuration because we have to run it from outside of the VS command shell. We need to find the path to the msbuild binary manually like this code.
> 
> Josh, you are correct.  msbuild is not found in this case.
> When I build it locally, I have to add msbuild to my path.

I thought perhaps the build server would have resolved the path to msbuild.  

I'll work on adding the build path as you suggest.

angelortiz1007 commented:
> Oh, actually, I bet the problem is that it can't find msbuild in the arm64 build configuration because we have to run it from outside of the VS command shell. We need to find the path to the msbuild binary manually like this code.

So, msbuild.exe is found under:
C:\Program Files (x86)\Microsoft Visual Studio\2017\Enterprise\MSBuild\15.0\Bin or 
C:\Program Files (x86)\Microsoft Visual Studio\2017\Professional\MSBuild\15.0\Bin\amd64

I guess a solution would be something as follows:

            editions = ["Enterprise", "Professional", "Community", "BuildTools"]
            prog_files = os.environ.get("ProgramFiles(x86)")
            # C:\Program Files (x86)\Microsoft Visual Studio\2017\Enterprise\MSBuild\15.0\Bin
            # C:\Program Files (x86)\Microsoft Visual Studio\2017\Professional\MSBuild\15.0\Bin\amd64
            base_vs_path = os.path.join(prog_files, "Microsoft Visual Studio", "2017")
            vs_version = "15.0"
            for edition in editions:
                msbuildinstalldir = os.path.join(base_vs_path, "MSBuild", vs_version, "Bin")
                vcinstalldir = os.path.join(base_vs_path, edition, "VC")
                print("***** vcinstalldir = {}",format(vcinstalldir))
                print("***** msbuildinstalldir = {}",format(msbuildinstalldir))
                if os.path.exists(vcinstalldir) and os.path.exists(msbuildinstalldir):
                    # Add msbuildinstalldir to path
                    env['PATH'] = env['PATH'] + os.path.abspath(msbuildinstalldir)
                    break
            else:
                print("Can't f

jdm commented:
@bors-servo try=windows

bors-servo commented:
:hourglass: Trying commit 87cb1bcad36489f41f262738065f17d0b5a56b3e with merge 625778590a352b57b0eda9e5810de4ea5e8dae7b...

bors-servo commented:
:broken_heart: Test failed - [status-taskcluster](https://tools.taskcluster.net/task-group-inspector/#/S9M5st-3QjCbQ2ruaEjG7g)

jdm commented:
@bors-servo retry

bors-servo commented:
:hourglass: Trying commit 87cb1bcad36489f41f262738065f17d0b5a56b3e with merge 77578b1ad83c04646da4a16bbdc0b17c1dac2ca5...

bors-servo commented:
:broken_heart: Test failed - [status-taskcluster](https://tools.taskcluster.net/task-group-inspector/#/YuQARNXwSAaLt-1M0Wj_6A)

jdm commented:
@bors-servo retry

bors-servo commented:
:hourglass: Trying commit 87cb1bcad36489f41f262738065f17d0b5a56b3e with merge 900617eaa11e958d9b28112ea28377b330e4df43...

bors-servo commented:
:broken_heart: Test failed - [status-taskcluster](https://tools.taskcluster.net/task-group-inspector/#/QVAA0Q2JQ26K4W6ZcciO-Q)

jdm commented:
@bors-servo retry

bors-servo commented:
:hourglass: Trying commit 87cb1bcad36489f41f262738065f17d0b5a56b3e with merge 3d066b054f3f4ac9628dffbacb7479f9538b3ab0...

bors-servo commented:
:broken_heart: Test failed - [status-taskcluster](https://tools.taskcluster.net/task-group-inspector/#/NHLFm3gqSY-2zbsQGCtZ7w)

jdm commented:
```
Microsoft (R) Build Engine version 15.9.21+g9802d43bc3 for .NET Framework
Copyright (C) Microsoft Corporation. All rights reserved.
Build started 8/20/2019 7:25:11 PM.
     1>Project "C:\Users\task_1566326421\repo\support\hololens\ServoApp.sln" on node 1 (default targets).
     1>ValidateSolutionConfiguration:
         Building solution configuration "Debug|arm64".
     1>Project "C:\Users\task_1566326421\repo\support\hololens\ServoApp.sln" (1) is building "C:\Users\task_1566326421\repo\support\hololens\ServoApp\ServoApp.vcxproj" (2) on node 1 (default targets).
     2>C:\Program Files (x86)\Microsoft Visual Studio\2017\BuildTools\Common7\IDE\VC\VCTargets\Microsoft.CppCommon.targets(2128,3): error MSB4019: The imported project "C:\Program Files (x86)\Microsoft Visual Studio\2017\BuildTools\MSBuild\Microsoft\WindowsXaml\v15.0\Microsoft.Windows.UI.Xaml.Cpp.targets" was not found. Confirm that the path in the <Import> declaration is correct, and that the file exists on disk. [C:\Users\task_1566326421\repo\support\hololens\ServoApp\ServoApp.vcxproj]
     2>Done Building Project "C:\Users\task_1566326421\repo\support\hololens\ServoApp\ServoApp.vcxproj" (default targets) -- FAILED.
     1>Done Building Project "C:\Users\task_1566326421\repo\support\hololens\ServoApp.sln" (default targets) -- FAILED.
Build FAILED.
       "C:\Users\task_1566326421\repo\support\hololens\ServoApp.sln" (default target) (1) ->
       "C:\Users\task_1566326421\repo\support\hololens\ServoApp\ServoApp.vcxproj" (default target) (2) ->
         C:\Program Files (x86)\Microsoft Visual Studio\2017\BuildTools\Common7\IDE\VC\VCTargets\Microsoft.CppCommon.targets(2128,3): error MSB4019: The imported project "C:\Program Files (x86)\Microsoft Visual Studio\2017\BuildTools\MSBuild\Microsoft\WindowsXaml\v15.0\Microsoft.Windows.UI.Xaml.Cpp.targets" was not found. Confirm that the path in the <Import> declaration is correct, and that the file exists on disk. [C:\Users\task_1566326421\repo\support\hololens\ServoApp\ServoApp.vcxproj]
    0 Warning(s)
    1 Error(s)
```

angelortiz1007 commented:
> Microsoft (R) Build Engine version 15.9.21+g9802d43bc3 for .NET Framework
> Copyright (C) Microsoft Corporation. All rights reserved.
> Build started 8/20/2019 7:25:11 PM.
>      1>Project "C:\Users\task_1566326421\repo\support\hololens\ServoApp.sln" on node 1 (default targets).
>      1>ValidateSolutionConfiguration:
>          Building solution configuration "Debug|arm64".
>      1>Project "C:\Users\task_1566326421\repo\support\hololens\ServoApp.sln" (1) is building "C:\Users\task_1566326421\repo\support\hololens\ServoApp\ServoApp.vcxproj" (2) on node 1 (default targets).
>      2>C:\Program Files (x86)\Microsoft Visual Studio\2017\BuildTools\Common7\IDE\VC\VCTargets\Microsoft.CppCommon.targets(2128,3): error MSB4019: The imported project "C:\Program Files (x86)\Microsoft Visual Studio\2017\BuildTools\MSBuild\Microsoft\WindowsXaml\v15.0\Microsoft.Windows.UI.Xaml.Cpp.targets" was not found. Confirm that the path in the <Import> declaration is correct, and that the file exists on disk. [C:\Users\task_1566326421\repo\support\hololens\ServoApp\ServoApp.vcxproj]
>      2>Done Building Project "C:\Users\task_1566326421\repo\support\hololens\ServoApp\ServoApp.vcxproj" (default targets) -- FAILED.
>      1>Done Building Project "C:\Users\task_1566326421\repo\support\hololens\ServoApp.sln" (default targets) -- FAILED.
> Build FAILED.
>        "C:\Users\task_1566326421\repo\support\hololens\ServoApp.sln" (default target) (1) ->
>        "C:\Users\task_1566326421\repo\support\hololens\ServoApp\ServoApp.vcxproj" (default target) (2) ->
>          C:\Program Files (x86)\Microsoft Visual Studio\2017\BuildTools\Common7\IDE\VC\VCTargets\Microsoft.CppCommon.targets(2128,3): error MSB4019: The imported project "C:\Program Files (x86)\Microsoft Visual Studio\2017\BuildTools\MSBuild\Microsoft\WindowsXaml\v15.0\Microsoft.Windows.UI.Xaml.Cpp.targets" was not found. Confirm that the path in the <Import> declaration is correct, and that the file exists on disk. [C:\Users\task_1566326421\repo\support\hololens\ServoApp\ServoApp.vcxproj]
>     0 Warning(s)
>     1 Error(s)

Ref: https://stackoverflow.com/questions/45041972/microsoft-windows-ui-xaml-csharp-targets-missing-on-ci-server 

This is interesting.  Seems like the msbuild within the BuildTools directory does not have the missing file. If you do as the link above suggest it may work.

I looked in my development system and found the missing file under:
C:\Program Files (x86)\Microsoft Visual Studio\2017\Professional\MSBuild\Microsoft\WindowsXaml\v15.0

jdm commented:
Yeah, I think it's another missing component in https://github.com/servo/servo/blob/17f423723c0206c934f104eb3352f39a9533aa41/etc/taskcluster/windows/first-boot.ps1#L77-L84. I'm rebuilding the CI image with it installed, hopefully.

jdm commented:
@bors-servo retry

bors-servo commented:
:hourglass: Trying commit 87cb1bcad36489f41f262738065f17d0b5a56b3e with merge 44e8d2f3a7bf731fae564b7d208d56f8a51f5719...

bors-servo commented:
:broken_heart: Test failed - [status-taskcluster](https://tools.taskcluster.net/task-group-inspector/#/KVbmB8k-RB-hTH8qDSCb4w)

jdm commented:
@bors-servo retry

bors-servo commented:
:hourglass: Trying commit 87cb1bcad36489f41f262738065f17d0b5a56b3e with merge b3a60cb179212d9f28ec1e80320ab9c128f3bef1...

bors-servo commented:
:broken_heart: Test failed - [status-taskcluster](https://tools.taskcluster.net/task-group-inspector/#/NuU8-3xyQM-X6WX2Hv8ixg)

jdm commented:
Progress!
```
Build started 8/20/2019 9:26:58 PM.
     1>Project "C:\Users\task_1566333818\repo\support\hololens\ServoApp.sln" on node 1 (default targets).
     1>ValidateSolutionConfiguration:
         Building solution configuration "Debug|arm64".
     1>Project "C:\Users\task_1566333818\repo\support\hololens\ServoApp.sln" (1) is building "C:\Users\task_1566333818\repo\support\hololens\ServoApp\ServoApp.vcxproj" (2) on node 1 (default targets).
     2>C:\Program Files (x86)\Microsoft Visual Studio\2017\BuildTools\Common7\IDE\VC\VCTargets\Microsoft.Cpp.WindowsSDK.targets(46,5): error MSB8036: The Windows SDK version 10.0.18362.0 was not found. Install the required version of Windows SDK or change the SDK version in the project property pages or by right-clicking the solution and selecting "Retarget solution". [C:\Users\task_1566333818\repo\support\hololens\ServoApp\ServoApp.vcxproj]
     2>Done Building Project "C:\Users\task_1566333818\repo\support\hololens\ServoApp\ServoApp.vcxproj" (default targets) -- FAILED.
     1>Done Building Project "C:\Users\task_1566333818\repo\support\hololens\ServoApp.sln" (default targets) -- FAILED.
Build FAILED.
```

angelortiz1007 commented:
> Progress!
> Build started 8/20/2019 9:26:58 PM.
>      1>Project "C:\Users\task_1566333818\repo\support\hololens\ServoApp.sln" on node 1 (default targets).
>      1>ValidateSolutionConfiguration:
>          Building solution configuration "Debug|arm64".
>      1>Project "C:\Users\task_1566333818\repo\support\hololens\ServoApp.sln" (1) is building "C:\Users\task_1566333818\repo\support\hololens\ServoApp\ServoApp.vcxproj" (2) on node 1 (default targets).
>      2>C:\Program Files (x86)\Microsoft Visual Studio\2017\BuildTools\Common7\IDE\VC\VCTargets\Microsoft.Cpp.WindowsSDK.targets(46,5): error MSB8036: The Windows SDK version 10.0.18362.0 was not found. Install the required version of Windows SDK or change the SDK version in the project property pages or by right-clicking the solution and selecting "Retarget solution". [C:\Users\task_1566333818\repo\support\hololens\ServoApp\ServoApp.vcxproj]
>      2>Done Building Project "C:\Users\task_1566333818\repo\support\hololens\ServoApp\ServoApp.vcxproj" (default targets) -- FAILED.
>      1>Done Building Project "C:\Users\task_1566333818\repo\support\hololens\ServoApp.sln" (default targets) -- FAILED.
> Build FAILED.

This is Progress!!!!

Simple fix to add the SDK.  Will probably take about 1 hour to download and install :(

jdm commented:
There is a technique to add a Windows SDK as a NuGet package. I'm going to give that a shot.

jdm committed:
Add UWP build tools to CI images.

jdm commented:
@bors-servo try=windows

bors-servo commented:
:hourglass: Trying commit 1c2e226724afde4e65bb697ec9510e91e501f990 with merge 1739383e2dd90ddd2c6ee19ef3c25d9fbe68bdf1...

bors-servo commented:
:broken_heart: Test failed - [status-taskcluster](https://tools.taskcluster.net/task-group-inspector/#/RS5diYVoROOA1KKfKHcBlQ)

bors-servo commented:
:umbrella: The latest upstream changes (presumably #24008) made this pull request unmergeable. Please resolve the merge conflicts.

jdm commented:
The technique that my Microsoft contact recommended for adding the Windows SDK as a NuGet package is internal-only. I'm waiting to hear from my contact if there's an alternate solution possible.

angelortiz1007 commented:
> The technique that my Microsoft contact recommended for adding the Windows SDK as a NuGet package is internal-only. I'm waiting to hear from my contact if there's an alternate solution possible.

Too bad it is internal. 

BTW, when I installed it via the VS2007 installer it took a bit of time to download and install.

jdm commented:
Yeah, my backup plan is to downgrade the version of the SDK that we rely on to 10.0.17763.0 and install the Microsoft.VisualStudio.Component.Windows10SDK.17763 component along with the VS build tools as part of the Windows AMI creation process.

jdm committed:
Downgrade UWP app target platform.

jdm commented:
@bors-servo try=windows

bors-servo commented:
:hourglass: Trying commit c453e14fc56134280d4460f97fed625102493b0a with merge 7e79896fd0fc604e1d500d234a0fb78ad1e28fa6...

bors-servo commented:
:broken_heart: Test failed - [status-taskcluster](https://tools.taskcluster.net/task-group-inspector/#/cSWB0FnLSPCVN2J4a1Hsdw)

jdm commented:
Woohoo, both UWP builds we care about passed. This should be mergeable after #24025.

angelortiz1007 commented:
> Woohoo, both UWP builds we care about passed. This should be mergeable after #24025.

Outstanding!!!

bors-servo commented:
:umbrella: The latest upstream changes (presumably #24025) made this pull request unmergeable. Please resolve the merge conflicts.

jdm commented:
@bors-servo r+

bors-servo commented:
:pushpin: Commit c98878f has been approved by `jdm`

<!-- @bors-servo r=jdm c98878f8337bc3a4e2596ece03c46692d6f2ff82 -->

bors-servo commented:
:hourglass: Testing commit c98878f8337bc3a4e2596ece03c46692d6f2ff82 with merge 5dad49ee8dd0a6ceacc46d94fd1b426293a31a0a...

bors-servo commented:
:broken_heart: Test failed - [status-taskcluster](https://tools.taskcluster.net/task-group-inspector/#/RV8bImtwSd6v3UHuusJebw)

CYBAI commented:
```
Build started 8/22/2019 3:41:12 AM.
     1>Project "C:\Users\task_1566430230\repo\support\hololens\ServoApp.sln" on node 1 (default targets).
     1>ValidateSolutionConfiguration:
         Building solution configuration "Debug|x64".
     1>Project "C:\Users\task_1566430230\repo\support\hololens\ServoApp.sln" (1) is building "C:\Users\task_1566430230\repo\support\hololens\ServoApp\ServoApp.vcxproj" (2) on node 1 (default targets).
     2>C:\Users\task_1566430230\repo\support\hololens\ServoApp\ServoApp.vcxproj(323,5): error : This project references NuGet package(s) that are missing on this computer. Use NuGet Package Restore to download them.  For more information, see http://go.microsoft.com/fwlink/?LinkID=322105. The missing file is ..\packages\Microsoft.Windows.CppWinRT.2.0.190620.2\build\native\Microsoft.Windows.CppWinRT.props.
     2>Done Building Project "C:\Users\task_1566430230\repo\support\hololens\ServoApp\ServoApp.vcxproj" (default targets) -- FAILED.
     1>Done Building Project "C:\Users\task_1566430230\repo\support\hololens\ServoApp.sln" (default targets) -- FAILED.

Build FAILED.

       "C:\Users\task_1566430230\repo\support\hololens\ServoApp.sln" (default target) (1) ->
       "C:\Users\task_1566430230\repo\support\hololens\ServoApp\ServoApp.vcxproj" (default target) (2) ->
       (EnsureNuGetPackageBuildImports target) -> 
         C:\Users\task_1566430230\repo\support\hololens\ServoApp\ServoApp.vcxproj(323,5): error : This project references NuGet package(s) that are missing on this computer. Use NuGet Package Restore to download them.  For more information, see http://go.microsoft.com/fwlink/?LinkID=322105. The missing file is ..\packages\Microsoft.Windows.CppWinRT.2.0.190620.2\build\native\Microsoft.Windows.CppWinRT.props.

    0 Warning(s)
    1 Error(s)
```

jdm committed:
Remove explicit nuget setup in CI.

jdm commented:
@bors-servo try=windows

bors-servo commented:
:hourglass: Trying commit 57adaaaf1880ba00e3f9f0c5b75fbeb28c0f20a9 with merge 0d7280acf1403bb09e048a6067f18c6491dd40f2...

bors-servo commented:
:sunny: Test successful - [status-taskcluster](https://tools.taskcluster.net/task-group-inspector/#/GFOrfqWnQ9yjoxQ2PjabVA)
State: approved= try=True

jdm commented:
@bors-servo r+

bors-servo commented:
:pushpin: Commit 57adaaa has been approved by `jdm`

<!-- @bors-servo r=jdm 57adaaaf1880ba00e3f9f0c5b75fbeb28c0f20a9 -->

bors-servo commented:
:hourglass: Testing commit 57adaaaf1880ba00e3f9f0c5b75fbeb28c0f20a9 with merge 382ebcc98bd70b4cbb17c66d0c51483164711568...

bors-servo commented:
:broken_heart: Test failed - [status-taskcluster](https://tools.taskcluster.net/task-group-inspector/#/WkVu8H8ETkutfhoTrdbJ4Q)

jdm commented:
@bors-servo retry

bors-servo commented:
:bomb: Failed to start rebuilding: `Unknown error`

bors-servo commented:
:hourglass: Testing commit 57adaaaf1880ba00e3f9f0c5b75fbeb28c0f20a9 with merge ee9b3b3864ee7988d45e861c19b57b706160edcf...

bors-servo commented:
:sunny: Test successful - [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/13182), [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/13243), [status-taskcluster](https://tools.taskcluster.net/task-group-inspector/#/OIuVnlCrTECyK3WARZxiBg)
Approved by: jdm
Pushing ee9b3b3864ee7988d45e861c19b57b706160edcf to master...
